### PR TITLE
[epsonprojector] Add ESC/VP.net handshake for projectors with built-in ethernet

### PR DESCRIPTION
**Pull request by mlobstein**
Some newer projectors have a built-in Ethernet port that listens for a connection on port 3629. This change sends a special string to initialize the connection. After the initialization string is sent, the normal serial commands can be used. Without the string being sent, the projector would terminate the connection. A discovery service was also added to automatically locate the projector on the network and add the discovered thing to the inbox.

Previously raised in this issue: https://github.com/openhab/openhab1-addons/issues/5604

Test build:
https://github.com/mlobstein/mlobstein-beta-test/raw/master/org.openhab.binding.epsonprojector-3.0.0-SNAPSHOT.jar
https://github.com/mlobstein/mlobstein-beta-test/raw/master/org.openhab.binding.epsonprojector-3.0.0-SNAPSHOT-sources.jar

### CODE REVIEW

File: bundles/org.openhab.binding.epsonprojector/src/main/java/org/openhab/binding/epsonprojector/internal/discovery/EpsonProjectorDiscoveryService.java [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+                    properties.put(THING_PROPERTY_PORT, DEFAULT_PORT);
+
+                    logger.trace("Projector with UID {} discovered at IP: {}", epsonMulticastListener.getUID(),
+                            epsonMulticastListener.getIPAddress());
+
+                    ThingUID uid = new ThingUID(THING_TYPE_PROJECTOR_TCP, epsonMulticastListener.getUID());
+                    logger.trace("Creating epson projector discovery result for: {}, IP={}", uid,
+                            epsonMulticastListener.getIPAddress());
+                    thingDiscovered(DiscoveryResultBuilder.create(uid).withProperties(properties)
+                            .withLabel("Epson Projector " + epsonMulticastListener.getUID()).build());
```
Comment (fwolter) [inline bundles/org.openhab.binding.epsonprojector/src/main/java/org/openhab/binding/epsonprojector/internal/discovery/EpsonProjectorDiscoveryService.java:166]: Can you add the representation property?

File: bundles/org.openhab.binding.epsonprojector/src/main/java/org/openhab/binding/epsonprojector/internal/discovery/MulticastListener.java [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+    // Epson-specific properties defined in this binding
+    private String uid = "";
+    private String ipAddress = "";
+
+    // Epson projector devices announce themselves on a multicast port
+    private static final String EPSON_MULTICAST_GROUP = "239.255.250.250";
+    private static final int EPSON_MULTICAST_PORT = 9131;
+
+    // How long to wait in milliseconds for a discovery beacon
+    public static final int DEFAULT_SOCKET_TIMEOUT = 3000;
```
Comment (fwolter) [inline bundles/org.openhab.binding.epsonprojector/src/main/java/org/openhab/binding/epsonprojector/internal/discovery/MulticastListener.java:49]: It's good practice to append the unit to the field name e.g. TIMEOUT_SEC.

File: bundles/org.openhab.binding.epsonprojector/src/main/java/org/openhab/binding/epsonprojector/internal/discovery/MulticastListener.java [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+    }
+
+    /*
+     * Parse the announcement beacon into the elements needed to create the thing.
+     *
+     * Example Epson beacon:
+     * AMXB<-UUID=000048746B33><-SDKClass=VideoProjector><-GUID=EPSON_EMP001><-Revision=1.0.0>
+     */
+    private void parseAnnouncementBeacon(DatagramPacket packet) {
+        String beacon = (new String(packet.getData())).trim();
```
Comment (fwolter) [inline bundles/org.openhab.binding.epsonprojector/src/main/java/org/openhab/binding/epsonprojector/internal/discovery/MulticastListener.java:106]: Can you specify the encoding you expect? E.g. StandardCharsets.UTF_8

**Review by fwolter — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.binding.epsonprojector/src/main/java/org/openhab/binding/epsonprojector/internal/connector/EpsonProjectorTcpConnector.java [hunk 4]
```diff
@@ -58,6 +59,16 @@ public void connect() throws EpsonProjectorException {
         } catch (IOException e) {
             throw new EpsonProjectorException(e);
         }
+
+        // Projectors with built in Ethernet listen on 3629, we must send the handshake to initialize the connection
+        if (port == 3629) {
```
[the remark this answers is not in the document]
Comment (mlobstein) [reply bundles/org.openhab.binding.epsonprojector/src/main/java/org/openhab/binding/epsonprojector/internal/connector/EpsonProjectorTcpConnector.java:64]: done

File: bundles/org.openhab.binding.epsonprojector/src/main/java/org/openhab/binding/epsonprojector/internal/discovery/EpsonProjectorDiscoveryService.java [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+ *
+ * @author Mark Hilbush - Initial contribution
+ * @author Michael Lobstein - Adapted for the Epson Projector binding
+ */
+@NonNullByDefault
+@Component(service = DiscoveryService.class, immediate = true, configurationPid = "discovery.epsonprojector")
+public class EpsonProjectorDiscoveryService extends AbstractDiscoveryService {
+    private final Logger logger = LoggerFactory.getLogger(EpsonProjectorDiscoveryService.class);
+
+    private ScheduledExecutorService scheduledExecutorService = Executors.newScheduledThreadPool(1);
```
[the remark this answers is not in the document]
Comment (mlobstein) [reply bundles/org.openhab.binding.epsonprojector/src/main/java/org/openhab/binding/epsonprojector/internal/discovery/EpsonProjectorDiscoveryService.java:53]: It was late at night while I was coding this ;)

File: bundles/org.openhab.binding.epsonprojector/src/main/java/org/openhab/binding/epsonprojector/internal/discovery/EpsonProjectorDiscoveryService.java (hunk 1 above)
Comment (mlobstein) [reply bundles/org.openhab.binding.epsonprojector/src/main/java/org/openhab/binding/epsonprojector/internal/discovery/EpsonProjectorDiscoveryService.java:166]: done

File: bundles/org.openhab.binding.epsonprojector/src/main/java/org/openhab/binding/epsonprojector/internal/discovery/EpsonProjectorDiscoveryService.java [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+    }
+
+    @Reference
+    protected void setNetworkAddressService(NetworkAddressService networkAddressService) {
+        this.networkAddressService = networkAddressService;
+    }
+
+    protected void unsetNetworkAddressService(NetworkAddressService networkAddressService) {
+        this.networkAddressService = null;
+    }
```
[the remark this answers is not in the document]
Comment (mlobstein) [reply bundles/org.openhab.binding.epsonprojector/src/main/java/org/openhab/binding/epsonprojector/internal/discovery/EpsonProjectorDiscoveryService.java:147]: done

File: bundles/org.openhab.binding.epsonprojector/src/main/java/org/openhab/binding/epsonprojector/internal/discovery/EpsonProjectorDiscoveryService.java [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+        logger.debug("Epson projector discovery service deactivated");
+        stopBackgroundDiscovery();
+        super.deactivate();
+    }
+
+    @Override
+    @Modified
+    protected void modified(@Nullable Map<String, Object> configProperties) {
+        super.modified(configProperties);
+    }
```
[the remark this answers is not in the document]
Comment (mlobstein) [reply bundles/org.openhab.binding.epsonprojector/src/main/java/org/openhab/binding/epsonprojector/internal/discovery/EpsonProjectorDiscoveryService.java]: I guess not, removed.

**Review by fwolter — APPROVED** [review]
LGTM

**Review by cpmeister — APPROVED** [review]

**Pull request merged**